src/seedu/addressbook/storage/StorageFile.java

```diff
@@ -9,140 +9,149 @@
+
+	/* Note: Note the use of nested classes below.
+	 * More info https://docs.oracle.com/javase/tutorial/java/javaOO/nested.html
+	 */
+
+	/**
+	 * Signals that the given file path does not fulfill the storage filepath constraints.
+	 */
+	public static class InvalidStorageFilePathException extends IllegalValueException {
+		public InvalidStorageFilePathException(String message) {
+			super(message);
+		}
+	}
+
+	/**
+	 * Signals that some error has occured while trying to convert and read/write data between the application
+	 * and the storage file.
+	 */
+	public static class StorageOperationException extends Exception {
+		public StorageOperationException(String message) {
+			super(message);
+		}
+	}
+	
+	/**
+	 * Signals that file has been deleted already even though the program is still running.
+	 */
+	public static class FileDeletionException extends Exception {
+		public FileDeletionException() {
+			super("File has been deleted already");
+		}
+	}
+
+
+	private final JAXBContext jaxbContext;
+
+	public final Path path;
+
+	/**
+	 * @throws InvalidStorageFilePathException if the default path is invalid
+	 */
+	public StorageFile() throws InvalidStorageFilePathException {
+		this(DEFAULT_STORAGE_FILEPATH);
+	}
+
+	/**
+	 * @throws InvalidStorageFilePathException if the given file path is invalid
+	 */
+	public StorageFile(String filePath) throws InvalidStorageFilePathException {
+		try {
+			jaxbContext = JAXBContext.newInstance(AdaptedAddressBook.class);
+		} catch (JAXBException jaxbe) {
+			throw new RuntimeException("jaxb initialisation error");
+		}
+
+		path = Paths.get(filePath);
+		if (!isValidPath(path)) {
+			throw new InvalidStorageFilePathException("Storage file should end with '.txt'");
+		}
+	}
+
+	/**
+	 * Returns true if the given path is acceptable as a storage file.
+	 * The file path is considered acceptable if it ends with '.txt'
+	 */
+	private static boolean isValidPath(Path filePath) {
+		return filePath.toString().endsWith(".txt");
+	}
+
+	/**
+	 * Saves all data to this storage file.
+	 *
+	 * @throws StorageOperationException if there were errors converting and/or storing data to file.
+	 */
+	public void save(AddressBook addressBook) throws StorageOperationException {
+
+		/* Note: Note the 'try with resource' statement below.
+		 * More info: https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html
+		 */
+		try (final Writer fileWriter =
+				new BufferedWriter(new FileWriter(path.toFile()))) {
+
+			final AdaptedAddressBook toSave = new AdaptedAddressBook(addressBook);
+			final Marshaller marshaller = jaxbContext.createMarshaller();
+			marshaller.setProperty(Marshaller.JAXB_FORMATTED_OUTPUT, true);
+			marshaller.marshal(toSave, fileWriter);
+
+		} catch (IOException ioe) {
+			throw new StorageOperationException("Error writing to file: " + path);
+		} catch (JAXBException jaxbe) {
+			throw new StorageOperationException("Error converting address book into storage format");
+		}
```

Comment:
Indentation problem here?. Our coding standard requires you to use 4 spaces instead of tabs. You can configure Eclipse to convert tabs to spaces automatically.